scheduler/pkg/store/experiment/db.go

```diff
@@ -107,6 +120,13 @@ func (edb *ExperimentDBManager) restore(
 					return err
 				}
 				experiment := CreateExperimentFromSnapshot(&snapshot)
+				if experiment.Deleted && item.ExpiresAt() == 0 {
+					ttl := deletedExperimentTTL
+					err = edb.save(experiment, &ttl)
```

Comment:
Ideally this logic of adding ttl for deleted resources could have been done as part of a migration process to a new version of the db or at least as part of the construction of our db object as a migration is not strictly required here.

Having it here (while logically correct) is not making the code legible as it is mixing the concern of restore (the purpose of this function) with the extra logic of adding ttl in case it doesnt exist.

Happy to discuss views here.

Reply:
Also we can perhaps utilise the timestamps on these records and for deleted resources more than 24 hours ago we can remove them, which will allow users to get the benefit from this change upon upgrade to the new version (as opposed to the next restart of the scheduler).

Reply:
To clarify some elements in case Niall goes looking for migration logic. There is no logic for migration, just the code enabling that (a previous PR from Sherif has introduced versions that are stored in the DB and could be checked to implement custom logic) -- atm the migration from "v1" to "v2" mostly consists of setting that version in the DB if it doesn't exist.

I briefly pondered along the same lines as Sherif while reviewing the PR, and I think he makes a good point here. My take was that we're not really changing the db format, just the way in which we're using it (with TTLs) so it's not a "migration" per se, just a new feature. However, I now agree it would be cleaner to separate concerns and have this as part of migration (actual migration code would exist in `new{Pipeline,Experiment}DBManager`).

Reply:
With respect to record timestamps, elements in the db have a `item.Version` key which is supposed to hold the timestamp of the transaction where the item gets committed. However, I don't know if there are guarantees about that timestamp being related to wall-clock time -- it might be now but that may also change in the future (something to check).

If it exists and it can be fetched reliably (i.e is already being set automatically by the transactions), it would be quite nice to have this cleanup as we load old databases in the new version.

Reply:
> I think we can do this even without changing current{Pipeline,Experiment}snapshotVersion, by calling adding a separate function that adds TTLs and calling that from new{Pipeline,Experiment}DBManager. However, it would cost us an additional iteration of the db (rather than piggy-backing on the one we already do in restore)

I think my worry is that this logic of dealing with ttl if not done as a part of a proper migration is going to stick with us indefinitely in `restore` or in `new{Pipeline,Experiment}DBManager`, and this might be problematic in the future:
- if done in `restore` we will always have to keep this piggy-backing logic, which can restrict the way we might want to do restore in the future.
- if done in `new{Pipeline,Experiment}DBManager` we will always have to read the db once on a new scheduler even if it is not required.

However if the db has an new explicit version 3 (meaning that we have dealt with ttl). Then we dont have to do anything and it is cleaner.

Reply:
I've seen the DB migration logic and my thinking was similar to @lc525, in that it's not a schema migration, but rather some backfilling. I was thinking that this could be the first step in implementing proper logic for the deletion of resources, which would also take into account the in-memory store. I think the benefit of having the logic outside of the migration step, is that it will run on every restart, as opposed to once. If resources are not updated again after they have been deleted, this logic will be redundant, so a once only migration should suffice. If this logic does become redundant, as expected, it can be removed. Also the current restore logic is convulated anyway, in that it's restoring the DB and filling the in-memory store.

Regarding the `item.Version` timestamps, I think it would be much better to explicitly set some field on the resource itself, for example `DeletedAt`. The time a resource was deleted at will be useful for managing the in-memory store too.

Reply:
>  If this logic does become redundant, as expected, it can be removed

My point is in fact that we dont know when this logic has became redundant as it depends on the state of the db, which is not well defined unless we update the version record in the db to a new one.

As I appreciate it is not a schema change then perhaps a full migration is not strictly required. I still think that we should increase the version in the db to 3 (to know that we dont have any long lived deleted resources anymore) and if we decide to keep the logic in `restore` then we can first check whether the version is < 3 (to make it explicit).

I suggest also to have a quick think about how this will interact with the in-memory as it might affect the implementation here. However as we have progressed as far, this can happen in a follow-up work as suggested in one of the earlier comments.

Reply:
@sakoush Indeed, increasing the version would give us the benefit of only running this when needed. I also agree now that we should do that. 

At the moment, `migrateToDBCurrentVersion` is only called when there's no version, and a comment suggests that further migration logic would still be triggered from `new{Pipeline,Experiment}DBManager` though. So perhaps some small refactoring is needed for the migration logic.

@driev I agree that longer term we want that `DeletedAt` field. However, what Sherif was referring to was that on this particular migration (from "v2" to "v3"), what will happen is that the old database will get loaded, and the deleted experiments will be re-added to the in-memory store (via the callbacks in `restore`), irrespective of whether they were already deleted for a long time or not. So the point here is whether, even for this migration, we could already detect that a deleted experiment loaded from the old database is quite old, and no longer add it to the in-memory store.

The question is whether we could do that based on `item.Version`.

Reply:
> The question is whether we could do that based on `item.Version`. Otherwise, we give the customers v2.8.4, their in-memory stores remain just as big as they are now, and only on the first scheduler restart they will get the cleanup (if TTL expired)

Yeah, that would definitely work.

Reply:
@driev Sorry to pile this up in this PR's comments, but if we started discussing...

There is also the issue of rollback, which should be tested when we make changes to the persistent store. 

Say, a customer has deployed our next release, new scheduler comes up and updates the database (perhaps in this imaginary world we did make some changes to the format, like adding the "DeletedAt" field or perhaps other types of changes). 

Now, something goes terribly wrong and a significant fault is discovered in that Core 2 release. 

Customer wants to roll back. It should be possible to do so (use the new database with the old scheduler code). This should "just work" because we're serialising protobufs in the db, but I'm just adding this here so that we're on the lookout for updates to our code which might break this.

Reply:
As mentioned offline - `item.Version` isn't useful. Proper migration logic is necessary at some point, but I don't think it's required here, as the schema isn't changing. I've added some "cleanup" go routines, to keep the in-memory stores in sync with the DB (roughly).